### PR TITLE
[mixcloud] Add support for new frontend

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement

---

### Description of your *pull request* and other information

The decryption part is refactored to use iterators, and therefore a compat function was added.

The code is mostly self explanatory.

### CODE REVIEW

**Comment by ishitatsuyuki** [comment]
@dstftw Friendly ping, is there anything that I should elaborate?

File: youtube_dl/extractor/mixcloud.py [hunk 1]
```diff
@@ -84,54 +69,103 @@ def _real_extract(self, url):
...
+                    'url': self._decrypt_xor_cipher(key, base64.b64decode(info_json['streamInfo']['url']))
+                },
+                {
+                    'format_id': 'hls',
+                    'url': self._decrypt_xor_cipher(key, base64.b64decode(info_json['streamInfo']['hlsUrl']))
+                },
+                {
+                    'format_id': 'dash',
+                    'url': self._decrypt_xor_cipher(key, base64.b64decode(info_json['streamInfo']['dashUrl']))
+                }
```
Comment (dstftw) [inline youtube_dl/extractor/mixcloud.py]: 1. 151-162 code duplication.
2. Missing format keys should not breaks extraction.
3. HLS and DASH should be extracted with appropriate `_extract_*` methods.
4. HLS and DASH extraction should be non fatal.

File: youtube_dl/extractor/mixcloud.py [hunk 2]
```diff
@@ -84,54 +69,103 @@ def _real_extract(self, url):
...
+                 r'(?:m|data)-tooltip=["\']([\d,.]+) plays'],
+                webpage, 'play count', default=None))
+
+        else:
+            title = info_json['name']
+            thumbnail = 'https://thumbnailer.mixcloud.com/unsafe/600x600/' + info_json['picture']['urlRoot']
+            uploader = info_json['owner']['displayName']
+            uploader_id = info_json['owner']['username']
+            description = info_json['description']
+            view_count = info_json['plays']
```
Comment (dstftw) [inline youtube_dl/extractor/mixcloud.py]: None of the optional fields should break extraction if missing.

File: youtube_dl/extractor/mixcloud.py [hunk 3]
```diff
@@ -84,54 +69,103 @@ def _real_extract(self, url):
...
+            # Decode
+            encrypted_play_info = base64.b64decode(encrypted_play_info)
+        else:
+            # New path
+            full_info_json = self._parse_json(self._html_search_regex(
+                r'<script id="relay-data" type="text/x-mixcloud">([^<]+)</script>', webpage, 'play info'), 'play info')
+            for item in full_info_json:
+                item_data = item.get("cloudcast", {}) \
+                    .get("data", {}) \
+                    .get("cloudcastLookup", {})
```
Comment (dstftw) [inline youtube_dl/extractor/mixcloud.py]: `try_get`, single quotes.

**Comment by kayb94** [comment]
Not sure if that is intented or not:
I just tried out your branch, ishitatsuyuki, with the media from #14088 and had to specify `format` as `normal` to get a listenable audio.
```

 > python -m youtube_dl https://www.mixcloud.com/aemidesign/mdma/ --list-formats[mixcloud] aemidesign-mdma: Downloading webpage
[mixcloud] aemidesign-mdma: Downloading webpage
[info] Available formats for aemidesign-mdma:
format code  extension  resolution note
normal       m4a        unknown    
hls          m3u8       unknown    
dash         mpd        unknown    (best)
 > python -m youtube_dl https://www.mixcloud.com/aemidesign/mdma/ -f normal
[mixcloud] aemidesign-mdma: Downloading webpage
[mixcloud] aemidesign-mdma: Downloading webpage
[download] Destination: MDMA-aemidesign-mdma.m4a
[download]  14.4% of 32.57MiB at  1.25MiB/s ETA 00:22^C
ERROR: Interrupted by user
```

Feel free to ignore, if I got something wrong here.

**Comment by ishitatsuyuki** [comment]
I'm using DASH with mpv without issues. Their server seems to terminate the connection weirdly though.

**Comment by ishitatsuyuki** [comment]
@kayb94 thanks for pointing out. That was indeed my mistake.

@dstftw I hope I addressed all your comments.

**Comment by kayb94** [comment]
Glad to be able to help!
Regards

**Comment by ishitatsuyuki** [comment]
@dstftw friendly ping, I think this is ready for another review!

File: youtube_dl/extractor/mixcloud.py [hunk 4]
```diff
@@ -84,54 +69,108 @@ def _real_extract(self, url):
...
+            }]
+
+            hls_encrypted = try_get(info_json, lambda x: x['streamInfo']['hlsUrl'])
+            if hls_encrypted is not None:
+                hls_url = self._decrypt_xor_cipher(key, base64.b64decode(hls_encrypted))
+                formats.extend(self._extract_m3u8_formats(hls_url, title))
+
+            dash_encrypted = try_get(info_json, lambda x: x['streamInfo']['dashUrl'])
+            if dash_encrypted is not None:
+                dash_url = self._decrypt_xor_cipher(key, base64.b64decode(dash_encrypted))
```
Comment (dstftw) [inline youtube_dl/extractor/mixcloud.py]: 157, 160-162, 165-167 code duplication.

File: youtube_dl/extractor/mixcloud.py [hunk 5]
```diff
@@ -84,54 +69,108 @@ def _real_extract(self, url):
...
+        if encrypted_play_info is not None:
+            # Decode
+            encrypted_play_info = base64.b64decode(encrypted_play_info)
+        else:
+            # New path
+            full_info_json = self._parse_json(self._html_search_regex(
+                r'<script id="relay-data" type="text/x-mixcloud">([^<]+)</script>', webpage, 'play info'), 'play info')
+            for item in full_info_json:
+                item_data = try_get(item, lambda x: x['cloudcast']['data']['cloudcastLookup'])
+                if try_get(item_data, lambda x: x['streamInfo']['url']) not in ['', None]:
```
Comment (dstftw) [inline youtube_dl/extractor/mixcloud.py]: `if try_get(item_data, lambda x: x['streamInfo']['url'])` is enough.

File: youtube_dl/extractor/mixcloud.py (hunk 4 above)
Comment (ishitatsuyuki) [reply youtube_dl/extractor/mixcloud.py]: What would you suggest then? The duplication is minimal.

**Comment by ishitatsuyuki** [comment]
@dstftw another review ping. If you have solutions to code duplication, please tell me how to do so.

File: youtube_dl/extractor/mixcloud.py (hunk 4 above)
Comment (dstftw) [reply youtube_dl/extractor/mixcloud.py]: Extracting duplicate code into a function obviously.

**Pull request closed without merging**